WIP: GPU basic rendering support, renderer lib

lpenguin opened the pull request:
## Preparing for `vange-rs` integration
### renderer - a separate library subproject for 2D and 3D rendering 
See: [Vangers/lib/renderer](https://github.com/KranX/Vangers/tree/02e93789459d8eb2bf39b6949bb11554ceb26f2e/lib/renderer). 

### Included libraries in the source tree
See [Vangers/lib/renderer/src/renderer/lib](https://github.com/KranX/Vangers/tree/02e93789459d8eb2bf39b6949bb11554ceb26f2e/lib/renderer/src/renderer/lib):
* [sokol_gfx](https://github.com/floooh/sokol) - a single-header 3D-API wrapper
* [HandmadeMath](https://github.com/HandmadeMath/Handmade-Math) - a single-header math library 

####
Additional dependences:
* OpenGL

### Main classes
* `renderer::core::AbstractCoreRenderer`: SDL_Renderer replacement
* `renderer::core::SokolCoreRenderer`: SDL_Renderer replacement implementation with [sokol_gfx](https://github.com/floooh/sokol) lib
* `renderer::scene::AbstractRenderer`: a class for future GPU rendering implementations
* `renderer::scene::SokolRenderer`: a simple GPU renderer with [sokol_gfx](https://github.com/floooh/sokol) lib
* `renderer::ResourceId<T>` a resource identifier which is basically a unique integral type. Each resource type mush typedef its own ResourceId. See `renderer::core::Texture` resource identifier as an example.   
* `renderer::core::Texture`: a texture identifier
* `renderer::scene::HeightMap`: a height map identifier
* `renderer::scene::util::MapUpdater`: an utility class for `HeightMap` updates

### xgraph changes
Since the `SDL_Renderer` is not working with OpenGL in parallel, I decided to create our own 2D rendering mini-framework and replace `SDL_Renderer` and `SDL_Texture` with `renderer::core::AbstractCoreRenderer` and `renderer::core::Texture`. Only theese two classes were replaced, we still use `SDL_Surface` for loading the BMP and pixel rendering.

### Original frame rendering changes (iGameMap, vrtMap)
Nothing much was changed: the software rendering into the `lineTcolor` is still here, but the actual screen rendering was replaced with the new AbstractRendering class. I didn't disable the software renderer completely, because it's hard to isolate and I decided to focus on `vange-rs` integration at first.

### Known problems
* Naming - the `renderer` word is not perfect, because it may conflict with variable names or headers. Consider more unique name like `xrender`?
* Naming - the namespace scheme is not perfect, the `renderer::core` namespace is confusing - it should be probably used for core types, so we should move 2D rendering classes to `renderer::2d` namespace or something
* The current unique integral type implementation `renderer::ResourceId<T>` is ugly, because it uses temporary enums for declaring the new type:
   ```c++
   template<typename T>
   struct ResourceId{ int32_t id; }; // omitted the member functions and consts for simplicity 

   enum class _HeightMapT; // temporary enum class
   typedef ResourceId<_HeightMapT> HeightMap; // uhhh
   ```
   We should consider a simpler approach. Like declaring each type explicitly: `struct HeightMap {int32_t id;}` or use `ResourceId` as base class without templates.
* Included `sokol_gfx` and `HandmadeMath` libraries directly into the code. We could get rid of `HandmadeMath` and use our own `3d_math.h` library. Need to decide how ship or install `sokol_gfx`.
* `renderer::scene::SokolRenderer` uses full texture updates instead of partial ones - the `sokol_gfx` library doesn't support partial updates unfortunately. It should not be a big problem since this class was created as dummy class, and the real GPU renderer should be `vange-rs` renderer.
* The `AbstractRenderer::render` function doesn't support camera rotation and seems to be bloated with too many arguments. I suggest we could create the `Camera` resource type and pass it to the `AbstractRenderer::render` 
   ```c++
   Camera camera_create_perspective(float FOV, float min_dist, float max_dist);
    // Transform is vec3 position and Quaternion rotation. I'm not sure how we implement this exactly
   void camera_set_transform(Camera camera, Transform transform);
   void camera_destroy(Camera camera);
   
   void render(int viewport_width, int viewport_height, Camera camera);
   ```

kvark commented on lib/renderer/src/renderer/ResourceStorage.h:
```diff
@@ -0,0 +1,86 @@
+#include "exception.h"
+
+namespace renderer {
+	template<typename TResourceId, typename TResource>
```
why is this `TResourceId` at all? Wouldn't it be the same as `Resource<TResource>`?

lpenguin replied:
Yes, it's just like that because I'm not totally happy with my current unique integral type approach. By *unique integral type* I mean that we have integer IDs, but we should not pass a Texture ID to the HeightMap methods for example. So I keep the ResourceStorage accepting any struct with `int32_t id` field in case if I change the current approach.

kvark replied:
We can cross this bridge when we get there. Seems strange to overcomplicate the API for something we don't know, at this point. Your choice though, not a big deal.

kvark commented on lib/renderer/src/renderer/ResourceStorage.h:
```diff
@@ -0,0 +1,86 @@
+			return it;
+		}
+
+		std::unordered_map<TResourceId, std::unique_ptr<TResource>, ResourceIdHash, ResourceIdEquals> _storage;
```
the unordered map is probably suboptimal here, and something like a freelist would turn out better
but I doubt it matters right now, nothing to worry about

lpenguin replied:
But would it be way less optimal for id lookup? I mean the idea of the ResourceStorage is to map resource id to the internal resource representation. For example if we have a Light resource and change the light position every frame, and the concrete implementation maps the light id to it's internal light object and updates the position

kvark replied:
yes, a freelist would mean that ID lookup is just indexing into an array. No hashing, bucketing, etc, involved.

lpenguin replied:
Ah, yes, exactly. I thought it's based on linked list

caiiiycuk commented:
+1 for renaming core namespace to 2d. Also little bit confusing about file layout, I prefer to group render files based on used library not by render type, so I mean put everything related to sokol under sokol root directory:
```
sokol/Sokol2dRenderer*
sokol/SokolRenderer*
```

Moreover I think we should try to move SokolRenderer as 3rd plugin library (so/dll), because this helps to understand which parts are still coupled with legacy renderer, and solve problem with dependencies on sokol-gfx, handmath. In current version, Sokol[2D|Abstract]Renderer is hard coupled with Soko[|3D]Renderer, you can't use Sokol3DRenderer without Sokol2DRenderer because sokolg-gfx initialization happens in Sokol2DRenderer.

So, probably we should remove Sokol2DRenderer from game core, and replace it with Dummy2DRenderer or SDL_renderer as before (?). When you load the renderer (so/dll) library you will get two pointer to 2DRenderer (can be null) and 3DRenderer and apply them to game. If library does not implement 2DRenderer (rust case) then Dummy/SDL_renderer is applied in other case 2d renderer from library will work.

I am not sure should we support mixing renderers like Sokol2D + Rust3D, I think we can assume that rendering library must provide 3d renderer and optionally can provide 2d renderer. So, in game core we will have hardcoded 2D renderer and no 3d renderers at all (only dummy one), this will make our live easier and we can no worry about mixing 3rd party libraries.

Another case that not covered by this PR is a switching between worlds, if we agreed that rust library will load resources by it's own then we need to notify renderer about which world is loaded, and probably we should wait until library says that world is ready.

I will continue to look PR next days, great work!

--
Also DummyRender is under renderer::scene::dummy, while Sokol is under renderer::scene.

kvark commented on lib/renderer/src/renderer/core/AbstractCoreRenderer.h:
```diff
@@ -0,0 +1,49 @@
+
+namespace renderer::core {
+	enum class TextureType {
+		RGBA32,
```
A bit of a strange name choice, even if it comes from SDL.
Most modern APIs would call it RGBA8

lpenguin replied:
Excatly, it was taken from SD :) I will change to RGBA8 then

kvark commented on lib/renderer/src/renderer/core/sokol/SokolTexture.cpp:
```diff
@@ -0,0 +1,391 @@
+			blend_state = {};
+			break;
+	}
+	sg_pipeline pip = sg_make_pipeline({
```
I find it peculiar that we are creating a pipeline as a part of the texture. Is this how people use Sokol generally?

lpenguin replied:
Uh, no that's just me copied the code from examples :)  I'll move it to the higher level

kvark commented on lib/renderer/src/renderer/core/sokol/SokolTexture.cpp:
```diff
@@ -0,0 +1,391 @@
+	render_context->bind.vertex_buffers[0] = vbuf;
+	render_context->vertex_buffer = vbuf;
+
+	uint16_t indices[] = {
```
if we are just drawing 2 triangles, can't we avoid index buffer altogether an just have a triangle strip with 4 vertices?

kvark commented on lib/renderer/src/renderer/core/sokol/SokolTexture.cpp:
```diff
@@ -0,0 +1,391 @@
+	std::unique_ptr<RenderContext> render_context = std::make_unique<RenderContext>(RenderContext{});
+
+	float vertices[] = {
+		-1.0f,  1.0f, 0.0f, 0.0f,
```
nit: we shouldn't really need this vertex buffer, the values can be produced from within the shader based on `gl_VertexID`

kvark commented on lib/renderer/src/renderer/scene/AbstractRenderer.h:
```diff
@@ -0,0 +1,35 @@
+		virtual HeightMap map_create(const MapDescription& map_description) = 0;
+		virtual void map_destroy(HeightMap map) = 0;
+		virtual void map_query(HeightMap map, int32_t* width, int32_t* height) = 0;
+		virtual void map_update_data(HeightMap map, const Rect& rect, uint8_t* height, uint8_t* meta) = 0;
```
where is the data to update with?

lpenguin replied:
The idea is that we copy a slice of data from the `lineT` to the `height` and `meta` and pass them to the `map_update_data`. And delete them after the call. So may be the arguments should be `const uint8_t*`. I wrote some comments to the AbstractRenderer.h, see my last commits and [MapUpdater](https://github.com/KranX/Vangers/blob/d933a9e034892247a497a354e9c7cf21c2b7fdab/lib/renderer/src/renderer/scene/util/MapUpdater.h) class

kvark replied:
ok, yes, these should be `const uint8_t*` then

kvark commented on lib/renderer/src/renderer/scene/sokol/HeightMap.cpp:
```diff
@@ -0,0 +1,376 @@
+}
+
+HeightMap::~HeightMap() {
+	delete[] height_map;
```
should also delete the meta and palette

kvark reviewed (commented):
Thank you for doing this grand rendering abstraction, and sorry for getting late to review this!
I believe most if not all of my notes are not important. The principle of abstraction makes sense, and we shouldn't hold it in the air until it's perfect. Instead, we should strike for iterating on it rapidly.

What do you see being the next step? Just trying to hook up vange-rs renderer and see if there are any roadblocks?

lpenguin commented:
Thanks for the review, guys!

@caiiiycuk 
>  Moreover I think we should try to move SokolRenderer as 3rd plugin library (so/dll)

Yes, that make sense

>+1 for renaming core namespace to 2d.

Ok, I'll think about the new naming scheme. Any ideas of the better name for the `renderer` library? I mean the current one is too broad :)

---

> So, probably we should remove Sokol2DRenderer from game core, and replace it with Dummy2DRenderer or SDL_renderer as before (?). When you load the renderer (so/dll) library you will get two pointers: to 2DRenderer (can be null), 3DRenderer and apply them to game. If library does not implement 2DRenderer (rust case) then Dummy/SDL_renderer is applied in other case 2d renderer from library will work.
> 
> I am not sure should we support mixing renderers like Sokol2D + Rust3D, I think we can assume that rendering library must provide 3d renderer and optionally can provide 2d renderer. So, in game core we will have hardcoded 2D renderer and no 3d renderers at all (only dummy one), this will make our live easier and we can no worry about mixing 3rd party libraries.

My main motivation of having the Sokol 2D renderer is that we cannot mix SDL_renderer and direct OpenGL calls (from SDL_GL_CreateContext). I have not found any information of using them in parallel (except of a weird SDL_GL_BindTexture call).  I suppose we cannot just use SDL_renderer and vange-rs renderer in parallel, for example. I guess SDL developers don't want anyone to mess with their internal SDL_Renderer context. So I decided that we should have our own OpenGL 2D renderer. 

My initial idea was that we **always** have Sokol 2D render as SDL_Renderer replacement and just pass a pointer of the OpenGL context to the concrete Abstract[Scene]Renderer implementation (to the vange-rs lib for example). The only problem here is that the concrete implementation may unintentionally mess with the internal Sokol OpenGL state. Uh, and sokol_gfx doesn't have a way to pass the OpenGL context to it, so the Sokol[Scene]Render just assumes that everything is already initialized..

I like your suggestion that every renderer implementation should provide a scene renderer and a 2D renderer. That could be another solution of the SDL_renderer/OpenGL problem. 

So I see a few possible options:
1. I'm wrong about using the SDL_renderer and OpenGL in parallel. I wish I could find a working example of that.
2. We permanently replace SDL_renderer with Sokol 2D renderer and pass it's opengl context to the Rust renderer. The OpenGL context comes from `SDL_GL_CreateContext` call.
3. We force Abstract[Scene]Renderer to implement both 2D and scene renderers. In that case we should implement 2D rendering on the Rust side
4. *Any other option?*


---

> Another case that not covered by this PR is a switching between worlds, if we agreed that rust library will load resources by it's own then we need to notify renderer about which world is loaded, and probably we should wait until library says that world is ready.

This is an important one: the current approach is to pass a world information from Vangers to the Abstract[Scene]Renderer implementation. See 
```c++
virtual void renderer::scene::AbstractRenderer::map_update_data(HeightMap map, const Rect& rect, uint8_t* height, uint8_t* meta) = 0; 
```
This method updates a part of the height map with new height and meta information. I plan to have the same scheme for mechous models and etc.  @kvark can we support such approach on the Rust side?

---

@kvark Thanks for  your notes :)
> The principle of abstraction makes sense, and we shouldn't hold it in the air until it's perfect. Instead, we should strike for iterating on it rapidly.

Totally agreed

> What do you see being the next step? Just trying to hook up vange-rs renderer and see if there are any roadblocks?

Yes, I think we should focus on rendering the height map in game using the vange-rs library. So we should decide what would be an interface of the vange-rs library and probably adapt the AbstractRenderer interface to it.

lpenguin commented:
Updates: 
* Removed HeightMap and Camera resource handles, since they are always single instance
* Added a stub for vange-rs implementation: https://github.com/KranX/Vangers/pull/517/commits/91882b2ba735772deb79ef8e00765b985e1792b8
* One can set a renderer instance here: https://github.com/KranX/Vangers/blob/40c518e601919d6c0e460b73999891a615f15bca/src/road.cpp#L381

kvark commented on lib/renderer/src/renderer/lib/sokol_gfx.h:
```diff
@@ -0,0 +1,15732 @@
+                u->offset = (uint16_t) cur_uniform_offset;
+                cur_uniform_offset += _sg_uniform_size(u->type, u->count);
+                if (u_desc->name) {
+                    u->gl_loc 
```
do we have to include the sokol header in-tree? could it be a git submodule pointing upstream instead?

lpenguin replied:
Yes, it should probably come from external source, but we don't use submodules here. May be i'll just delete sokol renderers because we don't get much profit from them now.

kvark commented on lib/renderer/src/renderer/scene/rust/vange_rs.h:
```diff
@@ -0,0 +1,82 @@
+
+
+// TODO: pointer?
+void vange_rs_camera_create(vange_rs_camera_description desc);
```
perhaps we can use a simpler prefix, like "rv_", which would stand for "Rusty Vangers"?

caiiiycuk replied:
Actually I think we should avoid using "rust" and "vange-rs" naming. As I think, we doing plugin system that allow us to add different types of renderers, "vange-rs" is one of them. But all of them share same interface. So, maybe we can call it "plugin", "external", "dynamic", etc.

caiiiycuk replied:
maybe dl_*

kvark replied:
What does "dl_" stand for?
I don't particularly care what the prefix is, honestly.
And I think "rv_" is fine, too. The whole effort is driven by the need to integrate Rusty Vangers rendering. If in the future we'll get more rendering plugins, not written in Rust, that's a very good place to be in, and I don't think "rv_" prefix is going to be a problem at that time.

caiiiycuk replied:
dynamic linking, same as dl library. I agree not big deal

kvark commented on lib/renderer/src/renderer/scene/rust/vange_rs.h:
```diff
@@ -0,0 +1,82 @@
+
+
+// TODO: pointer?
+void vange_rs_camera_create(vange_rs_camera_description desc);
```
it would be good to have a context object explicit, i.e.
```cpp
typedef void* rv_context;
rv_context rv_init();
rv_exit(rv_context);
rv_camera_init(rv_context, rv_camera_desc);
```

And so forth, every function having "rv_context" as the first argument.

kvark reviewed (commented):
Thank you for making those Rusty changes!

caiiiycuk commented on src/CMakeLists.txt:
```diff
@@ -91,10 +93,11 @@ TARGET_LINK_LIBRARIES(vangers
 	xtool
 	xgraph
 	xsound
+	dl
```
Is it really needed? We use static linkage (at least in rust part I seen staticlib crate)

lpenguin replied:
I should check this, but I remember that I get `symbol not found` on linux without `-ldl`

lpenguin commented:
Added [vange-rs](https://github.com/kvark/vange-rs) renderer via FFI.

How to build:
* install rustup
* clone https://github.com/kvark/vange-rs
* build `libvangers_ffi.a`
```
cd vange-rs/lib/ffi
cargo build
```
* copy `vange-rs/res` folder to vangers data dir
* Build Vangers: set `-DRUST_VANGERS_ROOT=<path>`, where `path` is directory containing `libvangers_ffi.a`

kvark commented on lib/renderer/src/renderer/scene/rust/vange_rs.h:
```diff
@@ -0,0 +1,101 @@
+#endif
+	extern int32_t rv_api_1;
+
+	rv_context rv_init(rv_init_descriptor desc);
```
there is no `rv_resize`?

lpenguin replied:
Yes, I didn't implement it yet, but I remember that I should add this one :)

lpenguin replied:
@kvark added `rv_api_1` check

kvark commented on lib/renderer/src/renderer/scene/rust/RustRenderer.cpp:
```diff
@@ -0,0 +1,192 @@
+#include <iostream>
+
```
would be nice to have a similar protection mechanism here that checks the API version like we do in rv_ffi

lpenguin replied:
Sure, I'll add it definitely

lpenguin commented on lib/renderer/src/renderer/ResourceStorage.h:
```diff
@@ -0,0 +1,86 @@
+		typedef typename std::unordered_map<TResourceId, std::unique_ptr<TResource>, ResourceIdHash, ResourceIdEquals>::iterator iterator;
+
+		ResourceStorage()
+			:_next(0)
```
I think we should never use the internal `id` field explicitly, so zero should be fine. Anyway I plan to refactor the "resource" system, since it's a bit over engineered..

lpenguin commented on lib/renderer/src/renderer/core/AbstractCoreRenderer.h:
```diff
@@ -0,0 +1,49 @@
+	};
+
+	enum class BlendMode {
+		None,
```
It means Copy, yes. Should it be called "Copy", what's the usual naming?

lpenguin commented on lib/renderer/src/renderer/scene/AbstractRenderer.h:
```diff
@@ -0,0 +1,81 @@
+		// Destroys the HeightMap
+		virtual void map_destroy() = 0;
+				
+		// Updates a rectangular area the selected HeightMap with height data from *height* and meta data from *meta*
```
It's an old comment, thanks. We use now `lineT` from MapDescription.

lpenguin commented on lib/renderer/src/renderer/scene/rust/vange_rs.h:
```diff
@@ -0,0 +1,82 @@
+#ifndef VANGE_RS_H
```
But they're already extern "C", see `#ifdef __cplusplus`

lpenguin commented on lib/renderer/src/renderer/scene/sokol/MapUpdater.cpp:
```diff
@@ -0,0 +1,81 @@
+
+		uint8_t* _lineT_r = lineT[iy + y_start];
+		if(_lineT_r != nullptr){
+			std::memcpy(height_map_r, _lineT_r + x_start, sizeof(uint8_t) * width);
```
I think one shouldn't use SokolRenderer now in general, because it was used only as simplified version for Rust renderer, like a stub before we've got the Rust renderer. 

The MapUpdater used an old approach, where I passed the map updates via region copies

lpenguin commented:
* Renamed `renderer::core::AbstractCoreRendere`r to `renderer::compositor::AbstractCompositor`
* Renamed `renderer::scene::AbstractRenderer` to `renderer::visualbackend::AbstractVisualBackend`
* Removed sokol_gfx related rendering stuff (`SokolRenderer`, `SokolCoreRenderer`)
* `rv_resize` support
* `rv_api_1` support

lpenguin commented on lib/renderer/src/renderer/visualbackend/rust/vange_rs.h:
```diff
@@ -0,0 +1,103 @@
+
+	void rv_map_update_data(rv_context context, rv_rect region);
+
+	void rv_map_update_palette(rv_context context, int32_t first_entry, int32_t entry_count, uint8_t* palette);
```
I'm thinking about passing `uint8_t** lineT` here, and remove it from `MapDescription`. This way, the rust side wouldn't keep the lineT pointer permanently.  @kvark what do you think?

kvark replied:
yep, that is fine

lpenguin commented:
@kvark i'm thinking about new name for the `libvangers_ffi` library, because it's a bit confusing: one can decide that it's related to C++ Vangers. Should we rename it to something like `librust_vangers_ffi` or `libvange_rs_ffi`?

kvark commented:
The name seems fine. We could change to something like `librusty_vangers` or even `librv_ffi`.

lpenguin commented:
@kvark `librusty_vangers`  looks good
